phabricatoremails/render.py

```diff
@@ -0,0 +1,420 @@
+                    )
+
+                for reviewer in body.reviewers:
+                    batch.target(reviewer.recipient, "updated", reviewer=reviewer)
+            elif kind == "revision-abandoned":
+                if is_secure:
+                    body = SecureRevisionAbandoned(
+                        [parse_recipient(reviewer) for reviewer in body["reviewers"]],
+                        body["commentCount"],
+                        body["transactionLink"],
+                    )
+                else:
+                    body = RevisionAbandoned(
+                        body.get("mainComment"),
+                        [
+                            parse_inline_comment(inline)
+                            for inline in body["inlineComments"]
+                        ],
+                        body["transactionLink"],
+                        [parse_recipient(reviewer) for reviewer in body["reviewers"]],
+                    )
+                batch.target_all(body.reviewers, "abandoned")
+            elif kind == "revision-reclaimed":
+                if is_secure:
+                    body = SecureRevisionReclaimed(
+                        [parse_recipient(reviewer) for reviewer in body["reviewers"]],
+                        body["commentCount"],
+                        body["transactionLink"],
+                    )
+                else:
+                    body = RevisionReclaimed(
+                        body.get("mainComment"),
+                        [
+                            parse_inline_comment(inline)
+                            for inline in body["inlineComments"]
+                        ],
+                        body["transactionLink"],
+                        [parse_recipient(reviewer) for reviewer in body["reviewers"]],
+                    )
+                batch.target_all(body.reviewers, "reclaimed")
+            elif kind == "revision-created":
+                if is_secure:
+                    body = SecureRevisionCreated(
+                        [parse_reviewer(reviewer) for reviewer in body["reviewers"]],
+                    )
+                else:
+                    body = RevisionCreated(
+                        [parse_affected_file(file) for file in body["affectedFiles"]],
+                        [parse_reviewer(reviewer) for reviewer in body["reviewers"]],
+                    )
+                for reviewer in body.reviewers:
+                    batch.target(reviewer.recipient, "created", reviewer=reviewer)
+            else:
+                raise ParseError("unexpected revision event kind: {}".format(kind))
+
+            actor_name = event["actorName"]
+
+            def query_thread(revision_id):
+                existing = (
+                    db.query(Thread).filter_by(phabricator_id=revision_id).first()
+                )
+                if existing:
+                    return existing
```

Comment:
> Whenever you use first, you should use order_by.

If you're querying on a table with a unique key, you're guaranteed to get either zero or one rows. Doesn't that make `order_by` redundant?

Reply:
> If you're querying on a table with a unique key, you're guaranteed to get either zero or one rows. Doesn't that make order_by redundant?

Ah I see -- It is unclear from this statement that you are expecting zero or one rows. Have you considered using `get`? It will return a single instance, or `None` if none exist. Otherwise, I would change this to use `one` or `one_or_none` to make it clear what you are expecting.

Reply:
Ooh, didn't know about `get()`, that sounds awesome when doing primary-key-based querying

Reply:
Hmm, `get(...)` only works for the primary key, which `phabricator_id` is not :(
Let me know if there's a cleaner way of getting 0 or 1 items by a unique key

Reply:
Sure, as mentioned
>  I would change this to use `one` or `one_or_none` to make it clear what you are expecting.